File: a800mon/inputwidget.py
import curses

from .actions import Actions
from .app import InputComponent, VisualComponent
from .appstate import state
from .ui import Color


class InputWidget(VisualComponent):
    def __init__(
        self,
        window,
        color: Color = Color.TEXT,
        max_length: int | None = None,
        on_enter=None,
        on_change=None,
    ):
        super().__init__(window)
        self.color = color
        self.max_length = max_length
        self.on_enter = on_enter
        self.on_change = on_change
        self._buffer = ""

    @property
    def buffer(self) -> str:
        return self._buffer

    def activate(self, initial_buffer: str = ""):
        self._buffer = ""
        self.set_buffer(initial_buffer)

    def deactivate(self):
        self._buffer = ""

    def _normalize_char(self, ch: str) -> str:
        return ch

    def _is_char_allowed(self, ch: str) -> bool:
        code = ord(ch)
        return 32 <= code <= 126

    def _to_value(self):
        return self._buffer

    def set_buffer(self, value: str) -> bool:
        text = str(value)
        if self.max_length is not None:
            text = text[: self.max_length]
        if text == self._buffer:
            return False
        self._buffer = text
        self.emit_change()
        return True

    def append_char(self, ch: int) -> bool:
        if ch < 0 or ch > 255:
            return False
        char = chr(ch)
        char = self._normalize_char(char)
        if not self._is_char_allowed(char):
            return False
        if self.max_length is not None and len(self._buffer) >= self.max_length:
            return False
        self._buffer += char
        self.emit_change()
        return True

    def backspace(self) -> bool:
        if not self._buffer:
            return False
        self._buffer = self._buffer[:-1]
        self.emit_change()
        return True

    def emit_change(self):
        if self.on_change is None:
            return
        self.on_change(self._to_value())

    def emit_enter(self):
        if self.on_enter is None:
            return
        self.on_enter(self._to_value())

    def render(self, force_redraw=False):
        self.window.cursor = 0, 0
        attr = self.color.attr() | curses.A_REVERSE
        text = state.input_buffer
        self.window.print(text, attr=attr)
        self.window.fill_to_eol(attr=attr)
        cursor_x = len(text)
        if cursor_x > self.window._iw - 1:
            cursor_x = self.window._iw - 1
        if cursor_x < 0:
            cursor_x = 0
        self.window.cursor = (cursor_x, 0)


class AddressInputWidget(InputWidget):
    def __init__(self, *args, **kwargs):
        kwargs["max_length"] = 4
        super().__init__(*args, **kwargs)

    def _normalize_char(self, ch: str) -> str:
        return ch.upper()

    def _is_char_allowed(self, ch: str) -> bool:
        return ("0" <= ch <= "9") or ("A" <= ch <= "F")

    def _to_value(self):
        if not self._buffer:
            return None
        return int(self._buffer, 16)

    def render(self, force_redraw=False):
        self.window.cursor = 0, 0
        attr = self.color.attr() | curses.A_REVERSE
        text = state.input_buffer[-4:].upper().rjust(4, "0")
        self.window.print(text, attr=attr)
        self.window.fill_to_eol(attr=attr)
        self.window.cursor = (4, 0)


class InputWidgetManager(InputComponent):
    def __init__(self, dispatcher, rebuild_screen):
        self._dispatcher = dispatcher
        self._rebuild_screen = rebuild_screen
        self._active_widget = None
        self._snapshot = ""
        self._replace_on_next_input = False

    def open(self, widget: InputWidget, initial_buffer: str):
        self._active_widget = widget
        self._snapshot = str(initial_buffer)
        self._replace_on_next_input = True
        widget.activate(initial_buffer)
        self._dispatcher.dispatch(Actions.SET_INPUT_BUFFER, widget.buffer)
        self._dispatcher.dispatch(Actions.SET_INPUT_FOCUS, True)
        widget.window.visible = True
        try:
            curses.curs_set(1)
        except curses.error:
            pass
        self._rebuild_screen()

    def _close(self):
        if self._active_widget is None:
            return
        self._dispatcher.dispatch(Actions.SET_INPUT_FOCUS, False)
        self._active_widget.window.visible = False
        self._active_widget.deactivate()
        self._active_widget = None
        self._snapshot = ""
        self._replace_on_next_input = False
        try:
            curses.curs_set(0)
        except curses.error:
            pass
        self._rebuild_screen()

    def _commit(self):
        if self._active_widget is None:
            return
        self._dispatcher.dispatch(Actions.SET_INPUT_BUFFER, self._active_widget.buffer)
        self._active_widget.emit_enter()
        self._close()

    def _cancel(self):
        if self._active_widget is None:
            return
        self._active_widget.set_buffer(self._snapshot)
        self._dispatcher.dispatch(Actions.SET_INPUT_BUFFER, self._snapshot)
        self._close()

    def handle_input(self, ch):
        if not state.input_focus or self._active_widget is None:
            return False

        if ch == 27:
            self._cancel()
            return True

        if ch in (10, 13, curses.KEY_ENTER):
            self._commit()
            return True

        if ch in (curses.KEY_BACKSPACE, 127, 8):
            self._replace_on_next_input = False
            if self._active_widget.backspace():
                self._dispatcher.dispatch(
                    Actions.SET_INPUT_BUFFER, self._active_widget.buffer
                )
            return True

        if self._replace_on_next_input:
            self._active_widget.set_buffer("")
            self._dispatcher.dispatch(Actions.SET_INPUT_BUFFER, "")
            self._replace_on_next_input = False

        if self._active_widget.append_char(ch):
            self._dispatcher.dispatch(
                Actions.SET_INPUT_BUFFER, self._active_widget.buffer
            )
            return True

        return True

File: go800mon/a800mon/inputwidget.go
package a800mon

import "context"

type InputWidget struct {
	BaseVisualComponent
	buffer    string
	maxLength int
	onChange  func(string)
	invalid   bool
}

func NewInputWidget(window *Window) *InputWidget {
	return &InputWidget{
		BaseVisualComponent: NewBaseVisualComponent(window),
		maxLength:           -1,
	}
}

func (i *InputWidget) Update(ctx context.Context) (bool, error) { return false, nil }
func (i *InputWidget) Render(force bool)                        {}
func (i *InputWidget) HandleInput(ch int) bool                  { return false }
func (i *InputWidget) PostRender(ctx context.Context) error     { return nil }

func (i *InputWidget) SetMaxLength(n int) {
	if n <= 0 {
		i.maxLength = -1
		return
	}
	i.maxLength = n
}

func (i *InputWidget) SetOnChange(cb func(string)) {
	i.onChange = cb
}

func (i *InputWidget) Buffer() string {
	return i.buffer
}

func (i *InputWidget) Activate(initial string) {
	_ = i.SetBuffer(initial)
}

func (i *InputWidget) Deactivate() {
	i.invalid = false
	_ = i.SetBuffer("")
}

func (i *InputWidget) SetBuffer(text string) bool {
	out := []rune(text)
	if i.maxLength > 0 && len(out) > i.maxLength {
		out = out[:i.maxLength]
	}
	normalized := string(out)
	if normalized == i.buffer {
		return false
	}
	i.buffer = normalized
	i.emitChange()
	return true
}

func (i *InputWidget) AppendChar(ch int) bool {
	if ch < 0 || ch > 255 {
		return false
	}
	r := rune(ch)
	if r < 32 || r > 126 {
		return false
	}
	if i.maxLength > 0 && len([]rune(i.buffer)) >= i.maxLength {
		return false
	}
	i.buffer += string(r)
	i.emitChange()
	return true
}

func (i *InputWidget) Backspace() bool {
	r := []rune(i.buffer)
	if len(r) == 0 {
		return false
	}
	i.buffer = string(r[:len(r)-1])
	i.emitChange()
	return true
}

func (i *InputWidget) SetInvalid(invalid bool) {
	i.invalid = invalid
}

func (i *InputWidget) Invalid() bool {
	return i.invalid
}

func (i *InputWidget) emitChange() {
	if i.onChange == nil {
		return
	}
	i.onChange(i.buffer)
}
